cylc/flow/prerequisite.py

```diff
@@ -115,16 +116,35 @@ def get_raw_conditional_expression(self):
 
     def set_condition(self, expr):
         """Set the conditional expression for this prerequisite.
-
         Resets the cached state (self._all_satisfied).
 
+        Examples:
+            # GH #3644 construct conditional expression when one task name
+            # is a substring of another: foo | xfoo => bar.
+            # Add 'foo' to the 'satisfied' dict before 'xfoo'.
+            >>> preq = Prerequisite(1)
+            >>> preq.satisfied = {
+            ...    ('foo', '1', 'succeeded'): False,
+            ...    ('xfoo', '1', 'succeeded'): False
+            ... }
+            >>> preq.set_condition("foo.1 succeeded|xfoo.1 succeeded")
+            >>> expr = preq.conditional_expression
+            >>> expr.split('|')  # doctest: +NORMALIZE_WHITESPACE
+            ['bool(self.satisfied[("foo", "1", "succeeded")])',
+            'bool(self.satisfied[("xfoo", "1", "succeeded")])']
+
         """
         self._all_satisfied = None
         if '|' in expr:
             # Make a Python expression so we can eval() the logic.
             for message in self.satisfied:
-                expr = expr.replace(self.MESSAGE_TEMPLATE % message,
-                                    self.SATISFIED_TEMPLATE % message)
+                # Use '\b' in case one task name is a substring of another
+                # and escape special chars ('.', timezone '+') in task IDs.
+                expr = re.sub(
+                    r"\b" + re.escape(self.MESSAGE_TEMPLATE % message) + r"\b",
```

Comment:
f-string for cylc-8.x branch only (7.8.x probably needs to stay as-is)?

```suggestion
                    fr"\b{re.escape(self.MESSAGE_TEMPLATE % message)}\b",
```

Reply:
Huh, I wasn't sure if I could nested templating like that!.

Reply:
But it does work, so I'll take it.